kernel/kernel-api/src/main/java/io/delta/kernel/internal/rowtracking/RowTrackingMetadataDomain.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.rowtracking;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.delta.kernel.internal.SnapshotImpl;
+import io.delta.kernel.internal.metadatadomain.JsonMetadataDomain;
+import java.util.Optional;
+
+/** Represents the metadata domain for row tracking. */
+public class RowTrackingMetadataDomain extends JsonMetadataDomain {
+
+  public static final String DOMAIN_NAME = "delta.rowTracking";
+
+  /** The highest assigned fresh row id for the table */
+  private long rowIdHighWaterMark;
+
+  /**
+   * Constructs a RowTrackingMetadataDomain with the specified row ID high water mark.
+   *
+   * @param rowIdHighWaterMark the row ID high water mark
+   */
+  @JsonCreator
+  public RowTrackingMetadataDomain(@JsonProperty("rowIdHighWaterMark") long rowIdHighWaterMark) {
+    this.rowIdHighWaterMark = rowIdHighWaterMark;
+  }
+
+  @Override
+  public String getDomainName() {
+    return DOMAIN_NAME;
+  }
+
+  public long getRowIdHighWaterMark() {
+    return rowIdHighWaterMark;
+  }
+
+  public void setRowIdHighWaterMark(long rowIdHighWaterMark) {
```

Comment:
any strong reason to make this class mutable as opposed to making all of its fields final and just creating a new instance?

It is typically best to minimize mutability. https://cs108.epfl.ch/archive/19/c/i/EffectiveJava_Item17.pdf

Reply:
Thanks for sharing the reference! I don't see reason for this class to be mutable, so I’ve updated it to be immutable.